### PR TITLE
Sub-PR to 846: Finish a solid draft for the roledb rewrite, and tests

### PR DESCRIPTION
**Pull request by adityasaky**
**Fixes issue #**: Part of the #660 fix, overall information at #846 

**Description of the changes being introduced by the pull request**: Fixes roledb tests (with some TODOs added to be discussed), based on changes introduced to roledb.

### CODE REVIEW

**Comment by awwad** [comment]
This is a PR to merge into the larger effort at #846 whatever changes were required to polish off [Step 2 as laid out here](https://github.com/theupdateframework/tuf/pull/846#issuecomment-479169228), using the tests in `test_roledb.py` to help guide you to do whatever else is required to have a coherent `roledb.py`.

This PR follows [a series of changes already made to roledb before this in 846](https://github.com/theupdateframework/tuf/pull/846/files/e6660cb5284adef868d420955dccb8b42388718f).

Further changes for roledb _might_ be necessary after this PR in order to finish #846, if we miss something, but the goal is to have a roledb that seems like it will support what the rest of TUF requires of it to finish #846.

In other words, with this PR merged, #846 should provide a roledb API that is complete, allowing other modules to query and populate role data, with convenience functions like `get_delegation_keyids` working.

**Comment by adityasaky** [comment]
Thanks for the summary, @awwad!

File: tuf/roledb.py [hunk 1]
```diff
@@ -900,10 +903,13 @@ def get_delegation_threshold(
...
-  # securesystemslib.exceptions.UnknownRoleError, or
-  # securesystemslib.exceptions.InvalidNameError.
-  _check_rolename(rolename, repository_name)
+  # Does 'rolename' have the correct object format?
+  # This check will ensure 'rolename' has the appropriate number of objects
+  # and object types, and that all dict keys are properly named.
+  tuf.formats.ROLENAME_SCHEMA.check_match(rolename)
+
+  # Raises securesystemslib.exceptions.InvalidNameError.
+  _validate_rolename(rolename)
```
Comment (awwad) [inline tuf/roledb.py:908]: A general tip -- NBD here -- that I'd give is to not overgeneralize comments explaining what something does.  In this case, while `check_match` can check the number of objects in something and their types, all it's doing here is making sure that rolename is a string -- it doesn't have objects of various types in it.  So it would be a little better to say that, or to leave it basic, like "check rolename's format".


Try to make commit messages a little more explicit about why a change is being made in the future.  In this case, the why is: the functions retrieve information from the delegating role, and it is not necessary for the delegated-to role to be loaded into roledb for us to retrieve the information, and not always reasonable to expect that the delegated-to role is loaded.

(As a totally unrelated matter, note that `_validate_rolename` should really be doing the `check_match` itself. When someone calls it, of course they're interested in that, too, and there's no reason to always call both. Has nothing to do with this PR.)

File: tests/test_roledb.py [hunk 2]
```diff
@@ -107,20 +109,23 @@ def test_clear_roledb(self):
     # Test for an empty roledb, a length of 1 after adding a key, and finally
     # an empty roledb after calling 'clear_roledb()'.
     self.assertEqual(0, len(tuf.roledb._roledb_dict['default']))
-    tuf.roledb._roledb_dict['default']['Root'] = {'keyids': ['123'], 'threshold': 1}
+    tuf.roledb._roledb_dict['default']['Root'] =\
+        securesystemslib.util.load_json_file(os.path.join(self.test_data_path,
+        "root.json"))['signed']
```
Comment (awwad) [inline tests/test_roledb.py:115]: Very minor style tips, for pedagogical purposes (NBD):
- space between `=` and `\`, per our coding guidelines, for readability
- Ideally, avoid `\` continuation.  It's reasonable here, since there aren't enough characters left to get to the `(`, and since adding a set of `()` might make it a tiny bit harder to read.

File: tests/test_roledb.py [hunk 3]
```diff
@@ -107,20 +109,23 @@ def test_clear_roledb(self):
...
     self.assertEqual(1, len(tuf.roledb._roledb_dict['default']))
     tuf.roledb.clear_roledb()
     self.assertEqual(0, len(tuf.roledb._roledb_dict['default']))
 
     # Verify that the roledb can be cleared for a non-default repository.
     rolename = 'targets'
-    roleinfo = {'keyids': ['123'], 'threshold': 1}
+    roleinfo = securesystemslib.util.load_json_file(os.path.join(self.test_data_path, "targets.json"))['signed']
```
Comment (awwad) [inline tests/test_roledb.py]: Style:
Definitely too long.  Break after one of the `(`, e.g.:

```
    roleinfo = securesystemslib.util.load_json_file(os.path.join(
        self.test_data_path, "targets.json"))['signed']
```

File: tuf/repository_tool.py [hunk 4]
```diff
@@ -1368,9 +1368,14 @@ def __init__(self, repository_name):
...
-                'signatures': [], 'version': 0, 'consistent_snapshot': False,
-                'expires': expiration, 'partial_loaded': False}
+    roleinfo = {"_type": "root",
+                "spec_version": "1.0",
+                "keys": {},
+                "roles": {},
+                "version": 1,
+                "consistent_snapshot": False,
+                "expires": "2030-01-01T00:00:00Z"}
+
```
Comment (awwad) [inline tuf/repository_tool.py:1374]: 1. Roles (including this blank one) should be built with `tuf.formats.build_dict_conforming_to_schema`.

2. This is just Root.  The adjustments for all the other metadata types aren't made.  With that said, does this change to `repository_tool` need to be in this `roledb` PR?  (I think no, but maybe there was some reason...?)  This sort of adjustment might be obviated by larger changes in `repository_tool`.  There might be one or two more test failures without this, perhaps (not sure), but this doesn't fix the underlying problem that's being left to the `repository_tool` PR anyway.

File: tuf/roledb.py [hunk 5]
```diff
@@ -1119,6 +1125,7 @@ def get_delegation(
   # Argument sanity check: top-level roles can only be delegated by root, and
   # delegated targets roles cannot be delegated by root.
   if top_level != (delegating_rolename == 'root'):
+    import pdb; pdb.set_trace()
```
Comment (awwad) [inline tuf/roledb.py]: stray debugger code

File: tests/test_roledb.py [hunk 6]
```diff
@@ -107,20 +109,23 @@ def test_clear_roledb(self):
...
 
     repository_name = 'example_repository'
     self.assertRaises(securesystemslib.exceptions.InvalidNameError, tuf.roledb.clear_roledb, repository_name)
     tuf.roledb.create_roledb(repository_name)
     tuf.roledb.add_role(rolename, roleinfo, repository_name)
-    self.assertEqual(roleinfo['keyids'], tuf.roledb.get_role_keyids(rolename, repository_name))
+    # TODO remove keyid check
+    # self.assertEqual(roleinfo['keyids'], tuf.roledb.get_delegation_keyids(rolename, repository_name))
```
Comment (awwad) [inline tests/test_roledb.py]: While this check is no longer needed, because that information is no longer included in `roleinfo`, some other check should be added in its place to confirm that there is now an entry in roledb.  Try `role_exists`.

File: tests/test_roledb.py [hunk 7]
```diff
@@ -138,7 +143,7 @@ def test_add_role(self):
     # Test conditions where the arguments are valid.
     self.assertEqual(0, len(tuf.roledb._roledb_dict['default']))
     rolename = 'targets'
-    roleinfo = {'keyids': ['123'], 'threshold': 1}
+    roleinfo = securesystemslib.util.load_json_file(os.path.join(self.test_data_path, "targets.json"))['signed']
```
Comment (awwad) [inline tests/test_roledb.py]: Style is less important in the tests, but still worth keeping to.  This is too long.  Break after a parenthesis, or the like.

File: tests/test_roledb.py [hunk 8]
```diff
@@ -190,7 +196,7 @@ def test_add_role(self):
   def test_role_exists(self):
     # Test conditions where the arguments are valid.
     rolename = 'targets'
-    roleinfo = {'keyids': ['123'], 'threshold': 1}
+    roleinfo = securesystemslib.util.load_json_file(os.path.join(self.test_data_path, "role1.json"))['signed']
```
Comment (awwad) [inline tests/test_roledb.py]: length

File: tests/test_roledb.py [hunk 9]
```diff
@@ -233,11 +239,8 @@ def test_remove_role(self):
     rolename = 'targets'
     rolename2 = 'release'
     rolename3 = 'django'
-    roleinfo = {'keyids': ['123'], 'threshold': 1}
-    roleinfo2 = {'keyids': ['123'], 'threshold': 1, 'delegations':
-      {'roles': [{'name': 'django', 'keyids': ['456'], 'threshold': 1}],
-       'keys': {'456': {'keytype': 'rsa', 'keyval': {'public': '456'}},
-      }}}
+    roleinfo = securesystemslib.util.load_json_file(os.path.join(self.test_data_path, "role1.json"))['signed']
+    roleinfo2 = securesystemslib.util.load_json_file(os.path.join(self.test_data_path, "role2.json"))['signed']
```
Comment (awwad) [inline tests/test_roledb.py]: length.  Please check other lines you've edited as well.  (Does your editor have the ability to draw a vertical line at 80 characters?)

File: tests/test_roledb.py [hunk 10]
```diff
@@ -53,6 +54,7 @@
 class TestRoledb(unittest.TestCase):
   def setUp(self):
     tuf.roledb.clear_roledb(clear_all=True)
+    self.test_data_path = os.path.join(os.getcwd(), "repository_data", "repository", "metadata")
```
Comment (awwad) [inline tests/test_roledb.py]: NBD, no real need to fix, but this is setting this field for the instance before every test.  It could just be a module-level variable for the test module, like `KEYS`, since it never needs to change.  (`KEYS` is pretty bad for other reasons, but that's not related to this PR. :) )

**Comment by awwad** [comment]
If your tests are mostly working, then I think you've probably not pushed all the roledb changes here.  I see a lot of errors, due to format issues and misnamed functions:

```bash
$ python test_roledb.py
EE.EEEEEEEEEERepository name does not exist: 'example_repository'
Repository name does not exist: 'example_repository'
.EEE
======================================================================
ERROR: test_add_role (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 143, in test_add_role
    self.assertEqual(None, tuf.roledb.add_role(rolename, roleinfo))
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_clear_roledb (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 122, in test_clear_roledb
    tuf.roledb.add_role(rolename, roleinfo, repository_name)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_create_roledb_from_root_metadata (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 563, in test_create_roledb_from_root_metadata
    self.assertEqual([keyid], tuf.roledb.get_role_keyids('root'))
AttributeError: module 'tuf.roledb' has no attribute 'get_role_keyids'

======================================================================
ERROR: test_get_delegated_rolenames (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 496, in test_get_delegated_rolenames
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_get_dirty_roles (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 671, in test_get_dirty_roles
    tuf.roledb.add_role(rolename, roleinfo1)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_get_role_info (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 318, in test_get_role_info
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_get_role_keyids (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 357, in test_get_role_keyids
    self.assertRaises(tuf.exceptions.UnknownRoleError, tuf.roledb.get_role_keyids, rolename)
AttributeError: module 'tuf.roledb' has no attribute 'get_role_keyids'

======================================================================
ERROR: test_get_role_paths (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 434, in test_get_role_paths
    self.assertRaises(tuf.exceptions.UnknownRoleError, tuf.roledb.get_role_paths, rolename)
AttributeError: module 'tuf.roledb' has no attribute 'get_role_paths'

======================================================================
ERROR: test_get_role_threshold (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 396, in test_get_role_threshold
    self.assertRaises(tuf.exceptions.UnknownRoleError, tuf.roledb.get_role_threshold, rolename)
AttributeError: module 'tuf.roledb' has no attribute 'get_role_threshold'

======================================================================
ERROR: test_get_rolenames (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 286, in test_get_rolenames
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_mark_dirty (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 703, in test_mark_dirty
    tuf.roledb.add_role(rolename, roleinfo1)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_remove_role (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 242, in test_remove_role
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_role_exists (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 197, in test_role_exists
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_unmark_dirty (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 726, in test_unmark_dirty
    tuf.roledb.add_role(rolename, roleinfo1)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

======================================================================
ERROR: test_update_roleinfo (__main__.TestRoledb)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_roledb.py", line 627, in test_update_roleinfo
    tuf.roledb.add_role(rolename, roleinfo)
  File "/Users/s/w/tuf/tuf/roledb.py", line 343, in add_role
    tuf.formats.ANYROLE_SCHEMA.check_match(roleinfo)
  File "/Users/s/.pyenv/versions/v366tt/lib/python3.6/site-packages/securesystemslib/schema.py", line 384, in check_match
    raise securesystemslib.exceptions.FormatError('Object did not match a'
securesystemslib.exceptions.FormatError: Object did not match a recognized alternative.

----------------------------------------------------------------------
Ran 17 tests in 0.014s

FAILED (errors=15)

**Review by awwad — CHANGES_REQUESTED** [review]
See comments above -- but primarily that the tests don't actually work.

File: tuf/repository_tool.py (hunk 4 above)
Comment (adityasaky) [reply tuf/repository_tool.py:1374]: This does belong in the repo_tool PR, and not here. This was one of my initial changes, one I made when I was still figuring things out. I apologise.

**Comment by awwad** [comment]
Merged into #846
